File: sos-app/apps/mobile-ios/SOSApp/Services/WebSocketManager.swift
import Foundation
import Starscream

class WebSocketManager: WebSocketDelegate {
    private var socket: WebSocket?
    private var emergencyId: String?

    var onLocationUpdate: ((LocationPoint) -> Void)?
    var onChatMessage: ((Message) -> Void)?
    var onAcknowledgment: ((Any) -> Void)?
    var onConnectionChange: ((Bool) -> Void)?

    func connect(emergencyId: String) {
        self.emergencyId = emergencyId

        guard let token = KeychainHelper.load(key: "accessToken") else {
            print("No access token found")
            return
        }

        var request = URLRequest(url: URL(string: "\(Config.wsURL)/socket.io/?EIO=4&transport=websocket")!)
        request.timeoutInterval = 5
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        socket = WebSocket(request: request)
        socket?.delegate = self
        socket?.connect()
    }

    func disconnect() {
        socket?.disconnect()
        socket = nil
    }

    func sendLocationUpdate(location: LocationPoint) {
        guard let emergencyId = emergencyId else { return }

        let payload: [String: Any] = [
            "emergencyId": emergencyId,
            "location": [
                "latitude": location.latitude,
                "longitude": location.longitude,
                "accuracy": location.accuracy,
                "timestamp": ISO8601DateFormatter().string(from: location.timestamp)
            ]
        ]

        sendEvent(event: "location-update", data: payload)
    }

    func sendChatMessage(text: String) {
        guard let emergencyId = emergencyId else { return }

        let payload: [String: Any] = [
            "emergencyId": emergencyId,
            "text": text,
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ]

        sendEvent(event: "chat-message", data: payload)
    }

    func sendTypingIndicator(isTyping: Bool) {
        guard let emergencyId = emergencyId else { return }

        let event = isTyping ? "typing-start" : "typing-stop"
        let payload: [String: Any] = ["roomId": emergencyId]

        sendEvent(event: event, data: payload)
    }

    private func sendEvent(event: String, data: [String: Any]) {
        guard let socket = socket, socket.isConnected else {
            print("Socket not connected")
            return
        }

        do {
            let jsonData = try JSONSerialization.data(withJSONObject: data)
            let message = "42[\"\(event)\",\(String(data: jsonData, encoding: .utf8) ?? "{}")]"
            socket.write(string: message)
        } catch {
            print("Failed to serialize message: \(error)")
        }
    }

    // MARK: - WebSocketDelegate

    func didReceive(event: WebSocketEvent, client: WebSocketClient) {
        switch event {
        case .connected(let headers):
            print("WebSocket connected: \(headers)")
            onConnectionChange?(true)

            // Join emergency room
            if let emergencyId = emergencyId {
                sendEvent(event: "join-room", data: ["roomId": emergencyId])
            }

        case .disconnected(let reason, let code):
            print("WebSocket disconnected: \(reason) with code: \(code)")
            onConnectionChange?(false)

        case .text(let text):
            handleMessage(text)

        case .binary(let data):
            print("Received binary data: \(data.count) bytes")

        case .ping:
            break

        case .pong:
            break

        case .viabilityChanged(let isViable):
            print("WebSocket viability changed: \(isViable)")

        case .reconnectSuggested(let shouldReconnect):
            if shouldReconnect {
                print("WebSocket reconnect suggested")
            }

        case .cancelled:
            print("WebSocket cancelled")
            onConnectionChange?(false)

        case .error(let error):
            print("WebSocket error: \(error?.localizedDescription ?? "Unknown error")")
            onConnectionChange?(false)

        case .peerClosed:
            print("WebSocket peer closed")
            onConnectionChange?(false)
        }
    }

    private func handleMessage(_ text: String) {
        // Socket.IO protocol parsing
        // Format: 42["event-name",{data}]

        guard text.hasPrefix("42[") else { return }

        let content = String(text.dropFirst(3).dropLast(1))
        let components = content.split(separator: ",", maxSplits: 1)

        guard components.count == 2 else { return }

        let eventName = String(components[0]).replacingOccurrences(of: "\"", with: "")
        let jsonString = String(components[1])

        guard let data = jsonString.data(using: .utf8) else { return }

        do {
            switch eventName {
            case "location-update":
                let locationData = try JSONDecoder().decode(LocationUpdateData.self, from: data)
                let location = LocationPoint(
                    latitude: locationData.location.latitude,
                    longitude: locationData.location.longitude,
                    accuracy: locationData.location.accuracy,
                    timestamp: ISO8601DateFormatter().date(from: locationData.location.timestamp) ?? Date()
                )
                onLocationUpdate?(location)

            case "chat-message":
                let message = try JSONDecoder().decode(Message.self, from: data)
                onChatMessage?(message)

            case "contact-acknowledged":
                let ack = try JSONDecoder().decode([String: Any].self, from: data)
                onAcknowledgment?(ack)

            default:
                print("Unknown event: \(eventName)")
            }
        } catch {
            print("Failed to decode message: \(error)")
        }
    }
}

// Helper structures
struct LocationUpdateData: Codable {
    let emergencyId: String
    let location: LocationData
}

struct LocationData: Codable {
    let latitude: Double
    let longitude: Double
    let accuracy: Double
    let timestamp: String
}

extension JSONDecoder {
    func decode<T>(_ type: [String: Any].Type, from data: Data) throws -> [String: Any] {
        guard let dictionary = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DecodingError.dataCorrupted(DecodingError.Context(
                codingPath: [],
                debugDescription: "Failed to decode dictionary"
            ))
        }
        return dictionary
    }
}

File: sos-app/apps/mobile-android/src/main/java/com/sosapp/android/data/remote/WebSocketManager.kt
package com.sosapp.android.data.remote

import com.sosapp.android.data.models.Location
import com.sosapp.android.data.models.Message
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.emitter.Emitter
import org.json.JSONObject
import java.net.URISyntaxException

class WebSocketManager {
    private var socket: Socket? = null
    private var emergencyId: String? = null

    var onLocationUpdate: ((Location) -> Unit)? = null
    var onChatMessage: ((Message) -> Unit)? = null
    var onConnectionChange: ((Boolean) -> Unit)? = null
    var onTypingStart: ((String) -> Unit)? = null
    var onTypingStop: ((String) -> Unit)? = null

    fun connect(emergencyId: String) {
        this.emergencyId = emergencyId

        try {
            val token = TokenManager.getAccessToken() ?: return

            val options = IO.Options().apply {
                query = "token=$token"
                transports = arrayOf("websocket", "polling")
            }

            socket = IO.socket("http://10.0.2.2:3000", options)

            socket?.apply {
                on(Socket.EVENT_CONNECT, onConnect)
                on(Socket.EVENT_DISCONNECT, onDisconnect)
                on(Socket.EVENT_CONNECT_ERROR, onConnectError)
                on("location-update", onLocationUpdateEvent)
                on("chat-message", onChatMessageEvent)
                on("typing-start", onTypingStartEvent)
                on("typing-stop", onTypingStopEvent)

                connect()
            }
        } catch (e: URISyntaxException) {
            e.printStackTrace()
        }
    }

    fun disconnect() {
        socket?.apply {
            off()
            disconnect()
        }
        socket = null
    }

    fun sendLocationUpdate(location: Location) {
        val json = JSONObject().apply {
            put("emergencyId", emergencyId)
            put("location", JSONObject().apply {
                put("latitude", location.latitude)
                put("longitude", location.longitude)
                put("accuracy", location.accuracy)
                put("timestamp", location.timestamp.time)
            })
        }

        socket?.emit("location-update", json)
    }

    fun sendChatMessage(text: String) {
        val json = JSONObject().apply {
            put("emergencyId", emergencyId)
            put("text", text)
            put("timestamp", System.currentTimeMillis())
        }

        socket?.emit("chat-message", json)
    }

    fun sendTypingIndicator(isTyping: Boolean) {
        val event = if (isTyping) "typing-start" else "typing-stop"
        val json = JSONObject().apply {
            put("roomId", emergencyId)
        }

        socket?.emit(event, json)
    }

    private val onConnect = Emitter.Listener {
        println("WebSocket connected")
        onConnectionChange?.invoke(true)

        // Join emergency room
        emergencyId?.let { id ->
            val json = JSONObject().apply {
                put("roomId", id)
            }
            socket?.emit("join-room", json)
        }
    }

    private val onDisconnect = Emitter.Listener {
        println("WebSocket disconnected")
        onConnectionChange?.invoke(false)
    }

    private val onConnectError = Emitter.Listener { args ->
        println("WebSocket connection error: ${args.firstOrNull()}")
        onConnectionChange?.invoke(false)
    }

    private val onLocationUpdateEvent = Emitter.Listener { args ->
        val data = args.firstOrNull() as? JSONObject ?: return@Listener

        try {
            val locationJson = data.getJSONObject("location")
            val location = Location(
                latitude = locationJson.getDouble("latitude"),
                longitude = locationJson.getDouble("longitude"),
                accuracy = locationJson.getDouble("accuracy"),
                timestamp = java.util.Date(locationJson.getLong("timestamp"))
            )

            onLocationUpdate?.invoke(location)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private val onChatMessageEvent = Emitter.Listener { args ->
        val data = args.firstOrNull() as? JSONObject ?: return@Listener

        try {
            val message = Message(
                id = data.getString("id"),
                emergencyId = data.getString("emergencyId"),
                senderId = data.getString("senderId"),
                senderName = data.optString("senderName"),
                text = data.getString("text"),
                createdAt = java.util.Date(data.getLong("createdAt")),
                deliveredAt = if (data.has("deliveredAt")) java.util.Date(data.getLong("deliveredAt")) else null,
                readAt = if (data.has("readAt")) java.util.Date(data.getLong("readAt")) else null
            )

            onChatMessage?.invoke(message)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private val onTypingStartEvent = Emitter.Listener { args ->
        val data = args.firstOrNull() as? JSONObject ?: return@Listener
        val userName = data.optString("userName", "Someone")
        onTypingStart?.invoke(userName)
    }

    private val onTypingStopEvent = Emitter.Listener { args ->
        val data = args.firstOrNull() as? JSONObject ?: return@Listener
        val userName = data.optString("userName", "Someone")
        onTypingStop?.invoke(userName)
    }
}
